opm/simulators/wells/MSWellHelpers.hpp

```diff
@@ -36,45 +37,47 @@ namespace Opm {
 
 namespace mswellhelpers
 {
-    // obtain y = D^-1 * x with a direct solver
+
+    /// Applies umfpack and checks for singularity
     template <typename MatrixType, typename VectorType>
     VectorType
-    invDXDirect(const MatrixType& D, VectorType x)
+    applyUMFPack(const MatrixType& D, std::shared_ptr<Dune::UMFPack<MatrixType> >& linsolver, VectorType x)
```

Comment:
Tangential to the topic of this PR I know, but what is the (typical) `VectorType` here?  Is there a reason we take the `x` by value?

Reply:
I think I tried changing that, but it seemed like UMFPack (in suitesparse) actually changed `x` when solving.

Reply:
Yes. It was tested before this way.

Reply:
> it seemed like UMFPack (in suitesparse) actually changed `x` when solving.

I won't dwell on this and I'm certainly not going to suggest we change it.  I will however remark that if `UMFPack` does change `x` then that's a violation of its declared contract&mdash;at least if `x` is the system right-hand side vector.

The `umfpack_*_solve()` functions are declared as
```C
/* From umfpack_solve.h, using "di" as the prototypical type example */
int umfpack_di_solve
(
    int sys,
    const int Ap [ ],
    const int Ai [ ],
    const double Ax [ ],
    double X [ ],
    const double B [ ],
    void *Numeric,
    const double Control [UMFPACK_CONTROL],
    double Info [UMFPACK_INFO]
);
```
with `B` denoting the right-hand side vector, `Ap`, `Ai`, and `Ax` being the coefficient matrix (CSC format) and `X` being the array in which to output solution to the linear system.

Reply:
Maybe it is how DUNE wrap it. 

```c++
334       double UMF_Apply_Info[UMFPACK_INFO];
335       Caller::solve(UMFPACK_A,
336                     umfpackMatrix_.getColStart(),
337                     umfpackMatrix_.getRowIndex(),
338                     umfpackMatrix_.getValues(),
339                     reinterpret_cast<double*>(&x[0]),
340                     reinterpret_cast<double*>(&b[0]),
341                     UMF_Numeric,
342                     UMF_Control,
343                     UMF_Apply_Info);
```

Maybe `reinterpret_cast<double*>(&b[0]),` should be `reinterpret_cast<const double*>(&b[0]),`?

Reply:
> Maybe it is how DUNE wrap it.

That may indeed be part of the reason here, yes.

> Maybe `reinterpret_cast<double*>(&b[0]),` should be `reinterpret_cast<const double*>(&b[0]),`?

I think that would be necessary, but maybe not quite sufficient.  I'd have to look into the details of the wrapper code to be sure.